### ISSUE TITLE
Add cache tag support to default Varnish config

### ISSUE THREAD

**Issue by geerlingguy** [geerlingguy/drupal-vm#397]
Documentation:

  - https://www.drupal.org/developing/api/8/cache/tags
  - https://www.varnish-cache.org/docs/trunk/users-guide/purging.html#bans
  - https://www.drupal.org/project/purge (w/ https://www.drupal.org/project/purge_purger_http for Varnish)
  - https://www.drupal.org/project/varnish (and [Drupal 8 port status](https://www.drupal.org/node/2365949))
  - http://foshttpcache.readthedocs.org/en/stable/varnish-configuration.html#tagging (might be simpler than that).

As I like to use Drupal VM as a good example of best practices for infrastructure configuration (e.g. the switch to PHP-FPM + Apache as a default), I'd also like to support D8 cache tags with Varnish out of the box, partly because I'm starting to build D8 sites that will rely on this functionality, and partly _just because_.

Also, for one good real-world use case of D8 + cache tags + Fastly in the wild, see: http://www.md-systems.ch/en/blog/2015-10/successful-relaunch-of-letemps-ch

**Comment by achton** [comment]
Indeed. We are also contemplating how to implement this in our hosting infrastructure in the best & simplest manner, and I had also looked at FOSHttpCache for pointers. Monitoring this closely, and will share if we come up with anything useful.

**Comment by achton** [comment]
@geerlingguy what is your plan here, if I may ask? Have you considered whether Purge module can help (the architecture is explained in the [3.0-beta1 release notes](https://www.drupal.org/node/2663370)), or are you going the way of wrapping around FosHttpCache? Or something else entirely?

Are you going to make use of the `http.response.debug_cacheability_headers: true` setting (now only recommended for development) to expose the `X-Drupal-Cache-Tags` header? Or how else do you intend on exposing the tags for Varnish?

And finally, which Varnish implementation to use? The Surrogate keys module for Varnish [has been open sourced](https://github.com/varnish/libvmod-xkey) and may do the trick, but there is also [the VCL snippet from FosHTTPCache](https://foshttpcache.readthedocs.org/en/1.4.1/varnish-configuration.html#tagging) which seems simpler.

**Comment by geerlingguy** [comment]
On the Drupal side, I'm thinking you'll want to use Purge to make sure cache tags are purged when appropriate.

On the Varnish side, I want to just make a VCL change. Don't want to add any additional modules or require any extra software/configuration. It should work OOTB with Varnish 4.x, I think. Just need to do the work to change over the default VCL that ships with Drupal VM.

Probably worthy of a blog post as well, to show people how they can start using cache tags with Varnish in Drupal 8 easily...

**Comment by geerlingguy** [comment]
Reading up as much as I can on various options and caveats with Bans, I'm also hoping to find a definitive answer to Wim's question from 2014: https://www.smashingmagazine.com/2014/04/cache-invalidation-strategies-with-varnish-cache/#comment-453091

Also, see:

  - https://www.varnish-cache.org/forum/topic/1128
  - https://www.drupal.org/node/2592471

**Comment by geerlingguy** [comment]
Some notes in getting Cache Tags working with purging:

  1. Set `http.response.debug_cacheability_headers: true` inside `services.yml`.
  2. Updates to the `drupalvm.vcl.j2` file (will be posted here soon) based on those in the [FOSHTTPCache library](http://foshttpcache.readthedocs.org/en/stable/varnish-configuration.html#tagging).

(Will be updated as I go.)

**Comment by geerlingguy** [comment]
See above PR.

**Comment by geerlingguy** [comment]
To set up purging support within Drupal:

  1. Install [purge_purger_http](https://www.drupal.org/project/purge_purger_http), which depends on [purge](https://www.drupal.org/project/purge): `drush en -y purge_purger_http`.
  2. TODO

**Comment by nielsvm** [comment]
I just pushed ``8.x-1.0-beta2`` at https://www.drupal.org/node/2692065 (still 403).

**Comment by nielsvm** [comment]
> Set http.response.debug_cacheability_headers: true inside services.yml (but see this comment/issue for more info on sending the right tags).

This shouldn't be necessary as ``purge`` takes care of this, though I need to change its implementation (but resulting behavior will stay the same).

**Comment by geerlingguy** [comment]
@nielsvm - Thanks for the clarification; doing some digging, I _did_ notice `purge.services.yml` sets the default:

```yml
parameters:

  # Enable Drupal's X-Drupal-Cache-Tags response header. This HTTP response
  # header describes what got rendered onto every page with descriptive text
  # keys uniquely identifying every rendered object. This then enables external
  # caching platforms - exactly what this module is for - to leverage this.
  http.response.debug_cacheability_headers: true
```

...however, if someone were to copy core's `default.services.yml` to `services.yml` in their site folder, then the default defined there (`http.response.debug_cacheability_headers: false`) will override the Purge module's default. So we definitely need to make it clear that _if the headers aren't being sent_, check and make sure you're not disabling the headers.

Or another option would be to do like the Fastly module and build in your own header (maybe even with a customizable name, defaulting to `X-Cache-Tags`?). Anyways, it looks like everything's working great now, with the latest beta releases of both modules; and no need for the Varnish module at all at this point!

**Comment by geerlingguy** [comment]
Working on a blog post with documentation for usage now.

**Comment by geerlingguy** [comment]
Blog post posted: http://www.jeffgeerling.com/blog/2016/use-drupal-8-cache-tags-varnish-and-purge

**Comment by nielsvm** [comment]
@geerlingguy thank you SO much for that article, very, very happy reading it and sharing it where I can. By no means is the ``purge_purger_http`` intended to be _the_ Varnish solution but in the meanwhile, this is all we got. I've been too busy to reach out to the ``varnish`` module maintainer for integration, so this module is the fallback solution.

In any case its safe to say that we're getting closer achieving our goal of perfect external cache invalidation on Drupal!

**Comment by wimleers** [comment]
>  I'm also hoping to find a definitive answer to Wim's question from 2014

I had cross-posted that comment to http://info.varnish-software.com/blog/advanced-cache-invalidation-strategies, but then they migrated their Drupal 7 site to a re-architected Drupal 7 site … _and lost all comments in the process!_ Whoever did that migration, _shame on you._

I have a cached copy, but sadly one that is from before my comment and Per Buer's reply: https://pinboard.in/cached/a31c4630dd2d/

I pinged @perbu on Twitter to ask him to restore those comments: https://twitter.com/wimleers/status/712573035132428288

But his answer was basically this (paraphrased from memory):

> Varnish can easily handle thousands of invalidations per second.

In other words: only at _enormous_ scale  — i.e. an invalidation frequency of >1000Hz — will you start to see problems. And then there is still `hashtwo`, which was open sourced for Varnish 4: `libvmod-xkey` — see https://www.varnish-cache.org/docs/trunk/whats-new/changes.html:

> libvmod-xkey: Secondary hash keys for cache objects, based on the hashtwo vmod written by Varnish Software. Allows for arbitrary grouping of objects to be purged in one go, avoiding use of ban invalidation. Also known as Cache Keys or Surrogate Key support.

**Comment by geerlingguy** [comment]
@wimleers - Thanks for the added context—that was my main worry, but in some simple testing I've been doing over the past few days, even with a few thousand bans in short succession, Varnish load barely showed any difference, and the ban lurker quickly cleaned out old requests.

I can see this technique scaling to most sites' traffic/content update levels quite easily. If you're a forum with hundreds of thousands of content updates per hour... then you'll probably be running into other issues before you have to worry about the lurker and optimize for libvmod-xkey. Though I'm going to try doing some testing on it as well, because it would be helpful to use in some situations!
